projects/components/src/filtering/filter/filter.service.ts

```diff
@@ -0,0 +1,205 @@
+    return this.navigationService.navigation$.pipe(map(() => this.getUrlFilters(attributes)));
+  }
+
+  public getUrlFilters(attributes: FilterAttribute[]): Filter[] {
+    return this.navigationService
+      .getAllValuesForQueryParameter(FilterService.FILTER_QUERY_PARAM)
+      .map(filterString => this.parseUrlFilterString(attributes, filterString))
+      .filter((filter): filter is Filter => filter !== undefined);
+  }
+
+  public setUrlFilters(filters: Filter[]): void {
+    this.navigationService.addQueryParametersToUrl({
+      [FilterService.FILTER_QUERY_PARAM]: filters.length === 0 ? undefined : filters.map(f => f.urlString)
+    });
+  }
+
+  public applyUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
+    const filters = this.getUrlFilters(attributes);
+
+    const foundIndex = filters.findIndex(f => filter.field === f.field);
+
+    if (foundIndex !== -1) {
+      filters[foundIndex] = filter;
+    } else {
+      filters.push(filter);
+    }
+
+    this.setUrlFilters([...filters]);
+  }
+
+  public removeUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
+    this.setUrlFilters([...this.getUrlFilters(attributes).filter(f => filter.field !== f.field)]);
+  }
+
+  private parseUrlFilterString(attributes: FilterAttribute[], filterString: string): Filter | undefined {
+    return attributes
+      .filter(attribute => this.filterBuilderLookupService.isBuildableType(attribute.type))
+      .flatMap(attribute => {
+        const filterBuilder = this.filterBuilderLookupService.lookup(attribute.type);
+        const supportedUrlOperators = filterBuilder.supportedOperators().map(toUrlFilterOperator);
+
+        const splitUrlFilter = AbstractFilterParser.splitFilterStringByOperator(
+          supportedUrlOperators,
+          filterString,
+          false
+        );
+
+        if (splitUrlFilter === undefined) {
+          return undefined;
+        }
+
+        const filterParser = this.filterParserLookupService.lookup(fromUrlFilterOperator(splitUrlFilter.operator));
+
+        const parsedFilter = filterParser.parseUrlFilterString(attribute, filterString);
+
+        if (parsedFilter === undefined) {
+          return undefined;
+        }
+
+        return splitUrlFilter.lhs === parsedFilter.field
+          ? filterBuilder.buildFilter(attribute, parsedFilter.operator, parsedFilter.value)
+          : undefined;
+      })
+      .find(splitFilter => splitFilter !== undefined);
+  }
+
+  /*
+   * Autocomplete
+   */
+
+  public autocompleteFilters(attributes: FilterAttribute[], text: string = ''): IncompleteFilter[] {
+    return attributes
+      .filter(attribute => this.filterBuilderLookupService.isBuildableType(attribute.type))
+      .flatMap(attribute => this.autocompleteFilter(attribute, text))
+      .filter(incompleteFilter => {
+        const isStringMatch = incompleteFilter.userString.toLowerCase().includes(text.toLowerCase());
```

Comment:
would autocomplete filter return filters that don't satisfy this condition? Assuming yes, otherwise the rest of the function is irrelevant - but based on the function names, would assume that this function is only mapping the singular version over each attribute

Reply:
Correct. This is a check on each attribute to see if it matches the current user string. As the user types, the attributes are matched and eliminated here if they don't include the user entered text.